### PR TITLE
Correct calculation issues in FindShapes by using code from Eigen

### PR DESCRIPTION
**Pull request by imikejackson**
Use Eigen to compute EigenValues and EigenVectors.
Remove all custom computation codes.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

### CODE REVIEW

File: Source/Plugins/Statistics/StatisticsFilters/FindShapes.cpp [hunk 1]
```diff
@@ -53,11 +55,90 @@
...
+ * @param b Second Value
+ * @param c Third Value
+ * @return The indices in their sorted order
+ */
+template <typename T>
+std::array<size_t, 3> TripletSort(T a, T b, T c, bool lowToHigh)
+{
+  const size_t A = 0;
+  const size_t B = 1;
+  const size_t C = 2;
```
Comment (JDuffeyBQ) [inline Source/Plugins/Statistics/StatisticsFilters/FindShapes.cpp]:
```suggestion
  constexpr size_t A = 0;
  constexpr size_t B = 1;
  constexpr size_t C = 2;
```

File: Source/Plugins/Statistics/StatisticsFilters/FindShapes.cpp [hunk 2]
```diff
@@ -53,11 +55,90 @@
...
+    // sorted[1] = b;
+    // sorted[2] = c;
+    idx = {A, B, C};
+  }
+
+  if(!lowToHigh)
+  {
+    size_t temp = idx[0];
+    idx[0] = idx[2];
+    idx[2] = temp;
```
Comment (JDuffeyBQ) [inline Source/Plugins/Statistics/StatisticsFilters/FindShapes.cpp]:
```suggestion
    std::swap(idx[0], idx[2]);
```

File: Source/Plugins/Statistics/StatisticsFilters/FindShapes.cpp [hunk 3]
```diff
@@ -338,23 +385,61 @@ void FindShapes::find_moments()
...
+    // clang-format off
+    //featureEigenVals =  Eigen::EigenSolver<Eigen::Matrix3f>::EigenvalueType eigenValues = es.eigenvalues();
+    Eigen::Matrix3f moment;
+    moment(0,0) = featureMoments[featureId * 6 + 0]; /**/ moment(0,1) = featureMoments[featureId * 6 + 3]; /**/moment(0,2) = featureMoments[featureId * 6 + 5];
+    moment(1,0) = featureMoments[featureId * 6 + 3]; /**/ moment(1,1) = featureMoments[featureId * 6 + 1]; /**/moment(1,2) = featureMoments[featureId * 6 + 4];
+    moment(2,0) = featureMoments[featureId * 6 + 5]; /**/ moment(2,1) = featureMoments[featureId * 6 + 4]; /**/moment(2,2) = featureMoments[featureId * 6 + 2];
+    Eigen::EigenSolver<Eigen::Matrix3f> es(moment);
+    Eigen::EigenSolver<Eigen::Matrix3f>::EigenvalueType eigenValues = es.eigenvalues();
+    Eigen::EigenSolver<Eigen::Matrix3f>::EigenvectorsType eigenVectors = es.eigenvectors();
+    // clang-format on
```
Comment (JDuffeyBQ) [inline Source/Plugins/Statistics/StatisticsFilters/FindShapes.cpp]:
```suggestion
    Eigen::Matrix3f moment;
    // clang-format off
    moment <<
      featureMoments[featureId * 6 + 0], featureMoments[featureId * 6 + 3], featureMoments[featureId * 6 + 5],
      featureMoments[featureId * 6 + 3], featureMoments[featureId * 6 + 1], featureMoments[featureId * 6 + 4],
      featureMoments[featureId * 6 + 5], featureMoments[featureId * 6 + 4], featureMoments[featureId * 6 + 2];
    // clang-format on
    Eigen::EigenSolver<Eigen::Matrix3f> es(moment);
    Eigen::EigenSolver<Eigen::Matrix3f>::EigenvalueType eigenValues = es.eigenvalues();
    Eigen::EigenSolver<Eigen::Matrix3f>::EigenvectorsType eigenVectors = es.eigenvectors();
```

We can use Eigen's comma initialization here.

File: Source/Plugins/Statistics/StatisticsFilters/FindShapes.cpp [hunk 4]
```diff
@@ -409,56 +499,56 @@ void FindShapes::find_moments2D()
...
-    m_FeatureMoments[i] = 0.0;
+    featureMoments[featureId] = 0.0;
   }
 
-  float x = 0.0f, y = 0.0f, x1 = 0.0f, x2 = 0.0f, y1 = 0.0f, y2 = 0.0f;
-  float xdist1 = 0.0f, xdist2 = 0.0f, xdist3 = 0.0f, xdist4 = 0.0f;
-  float ydist1 = 0.0f, ydist2 = 0.0f, ydist3 = 0.0f, ydist4 = 0.0f;
+  // float x = 0.0f, y = 0.0f, x1 = 0.0f, x2 = 0.0f, y1 = 0.0f, y2 = 0.0f;
+  //  float xdist1 = 0.0f, xdist2 = 0.0f, xdist3 = 0.0f, xdist4 = 0.0f;
+  //  float ydist1 = 0.0f, ydist2 = 0.0f, ydist3 = 0.0f, ydist4 = 0.0f;
```
Comment (JDuffeyBQ) [inline Source/Plugins/Statistics/StatisticsFilters/FindShapes.cpp]:
```suggestion
```

File: Source/Plugins/Statistics/StatisticsFilters/FindShapes.cpp [hunk 5]
```diff
@@ -467,96 +557,43 @@ void FindShapes::find_moments2D()
...
-    b = B / A;
-    b = sqrt(b) * a;
-    c = A / (a * a * a * b);
-
-    m_AxisLengths[3 * i] = static_cast<float>(a / m_ScaleFactor);
-    m_AxisLengths[3 * i + 1] = static_cast<float>(b / m_ScaleFactor);
-    m_AxisLengths[3 * i + 2] = static_cast<float>(c / m_ScaleFactor);
-    bovera = static_cast<float>(b / a);
-    covera = static_cast<float>(c / a);
+  const double multiplier = 1.0 / (4.0 * M_PI);
```
Comment (JDuffeyBQ) [inline Source/Plugins/Statistics/StatisticsFilters/FindShapes.cpp]:
```suggestion
  constexpr double multiplier = 1.0 / (4.0 * M_PI);
```

File: Source/Plugins/Statistics/StatisticsFilters/FindShapes.cpp [hunk 6]
```diff
@@ -53,11 +55,90 @@
...
+    // sorted[2] = b;
+    if(a > c)
+    {
+      // sorted[1] = a;
+      // sorted[0] = c;
+      idx = {C, A, B};
+    }
+    else
+    {
+      //  sorted[1] = c;
```
Comment (JDuffeyBQ) [inline Source/Plugins/Statistics/StatisticsFilters/FindShapes.cpp]:
```suggestion
      // sorted[1] = c;
```

File: Source/Plugins/Statistics/StatisticsFilters/FindShapes.cpp [hunk 7]
```diff
@@ -36,10 +36,12 @@
 #include "FindShapes.h"
 
 #include <array>
+#include <cmath>
```
Comment (JDuffeyBQ) [inline Source/Plugins/Statistics/StatisticsFilters/FindShapes.cpp:39]:
```suggestion
#include <cmath>
#include <utility>
```
For std::swap below.

**Review by JDuffeyBQ — CHANGES_REQUESTED** [review]

**Comment by imikejackson** [comment]
https://github.com/BlueQuartzSoftware/DREAM3DCI/pull/63

**Review by JDuffeyBQ — APPROVED** [review]

**Pull request merged**